Improve osu!mania playability on mobile devices

frenzibyte opened the pull request:
- [x] Requires testing on Android and ensuring `AndroidOrientationManager` works as it should (see https://github.com/ppy/osu/pull/31368#issuecomment-2566674159)
- [x] Depends on https://github.com/ppy/osu-framework/pull/6445 for iOS window API (would've split it out but the PR is already approved so)
- [x] Depends on https://github.com/ppy/osu-framework/pull/6477
- [x] Depends on framework bump
- Closes https://github.com/ppy/osu/issues/5331
- Addresses https://github.com/ppy/osu/discussions/16783


RFC. This is quite a packed PR, but the changes made here are closely tied to each other and cannot be separated as they cannot be merged / reviewed on their own (except for maybe the mania judgement positioning changes).

The aim of this PR is to lay down a path forward for supporting osu!mania gameplay on mobile devices and discuss the approach taken here and whether it's good / ethical / does not violate any existing gameplay rules or mechanics.

Preview of final product:

https://github.com/user-attachments/assets/7d0166c6-e9fc-4cdf-a0ef-66bb650f6c46

https://github.com/user-attachments/assets/a442484e-eb4f-462d-9717-54de67074dcb

### [Allow locking orientation on iOS in certain circumstances](https://github.com/ppy/osu/commit/7563a18c7fdcc40c33a1ef0e0ab5342ba8e879d1)

Adds support for making `OsuScreen`s able to force the device to a "portrait" orientation, to be used for mania-based `Player` screen sessions.

### [Improve osu!mania gameplay scaling on portrait orientation](https://github.com/ppy/osu/commit/5bdd7c2e8d8ce97be601667961bbc48e49ba08ab)

The main change of this PR. Naturally the entire game scene is surrounded by a scaling container that enforces the game to fit to the window size in a 4:3 aspect ratio. Since osu!mania is generally vertically based, this scaling behaviour is completely unfriendly on screens in portrait orientations.

To improve on this, I've added scaling logic in `ManiaPlayfieldAdjustmentContainer` that scales up the osu!mania stage just enough to be playable with a touchscreen on phones and tablet devices, and try to accommodate as much horizontal space as possible before scaling down on high key numbers / wide columns

Back-to-back comparison between master and this PR:

https://github.com/user-attachments/assets/d80fd975-1da9-4ef0-89ac-586e83921cb2

### [Make mania judgements relative to the hit target position](https://github.com/ppy/osu/commit/1e08b3dbdac1ed07fd56c0d55d83ce200053c336)

Following the commit above, since the playfield can be roughly scaled up to 125% its normal size, the judgements shift position, and they do so incorrectly, shifting down to below the hit target and look broken. This commit changes the anchoring of the judgements so they're tied to the hit target, so that when the playfield is scaled up, the judgements shift to the top rather than to the bottom.

Before:

https://github.com/user-attachments/assets/3b754516-bb2a-4a80-8d4d-954e7921ee3b

After:

https://github.com/user-attachments/assets/255fdf5b-84b7-407f-9171-28d94932e0ab

### [Replace `ManiaTouchInputArea` with touchable columns](https://github.com/ppy/osu/commit/5de48077686808cea4e19f907f94da937b48ea86)

Also following the scaling changes, which ensure that the playfield is scaled enough for the player to hit the notes with their fingers, I cannot find any use for `ManiaTouchInputArea` anymore, so I chose to remove it completely. This may be debatable, but I'm not sure refactoring it would make any sense and it's adding very bad visual clutter when displayed on top of the columns (it looks mispositioned from each column and basically broken to the user).

Due to this, I've looked into applying the portrait orientation changes to Android as well so osu!mania doesn't become broken there.

frenzibyte committed:
Allow locking orientation on iOS in certain circumstances

frenzibyte committed:
Improve osu!mania gameplay scaling on portrait orientation

frenzibyte committed:
Make mania judgements relative to the hit target position

This improves display in portrait screen, where the stage is scaled up.

frenzibyte committed:
Keep game in portrait mode when restarting

frenzibyte committed:
Abstractify orientation handling and add Android support

frenzibyte committed:
Replace `ManiaTouchInputArea` with touchable columns

Susko3 commented on osu.Game.Rulesets.Mania/UI/ManiaPlayfieldAdjustmentContainer.cs:
```diff
@@ -1,17 +1,64 @@
+            base.Update();
+
+            float aspectRatio = DrawWidth / DrawHeight;
+            bool isPortrait = aspectRatio < 4 / 3f;
```
Should use `osu.Framework.Utils.Precision`, or just compare with `1f`. Otherwise a 4:3 tablet in landscape might be considered portrait by float imprecision.

Another reason to use `1f` is that older 1280x1024 displays (5:4) are properly considered as landscape.

frenzibyte replied:
This didn't matter at first because I didn't want to apply a general upscale of the playfield, but after further considerations I've changed it as such. I'll change the boolean to check against `1f` instead.

frenzibyte replied:
https://github.com/ppy/osu/pull/31368/commits/64e557d00f98728e5a67d84c3158a8a11478c168

Susko3 reviewed (commented):
Haven't tested android yet. Do you have a set of tests that should be done? E.g. open screen X in landscape then rotate to portrait, is should/shouldn't rotate.

CI failures need attention.

frenzibyte commented:
Tests I've done:
 - On phone
   - Game is locked in landscape by default, can be left or right regardless of user rotation lock setting (or so is what I understand by "sensor")
   - Running osu! and in gameplay period, orientation is completely locked
   - Pausing unlocks orientation momentarily, resuming locks again.
   - Running osu!mania switches orientation to "portrait" and locks it there, nothing unlocks it until exiting.
   - Restarting does not reset back orientation to landscape, unless user exits.
 - On tablet
   - Game is allowed to rotate to any orientation by default, unless locked by auto-rotate lock.
   - Running osu! and in gameplay period, orientation is completely locked
   - Pausing unlocks orientation momentarily, resuming locks again.
   - Running osu!mania switches orientation to "portrait". When in gameplay period, orientation is locked, and when paused or in break you can switch orientation between portrait and upside-down portrait.

frenzibyte committed:
Simplify portrait check

PercyDan54 commented:
I get this behavior on my Xiaomi Pad
Not sure if it's some weird MIUI-specific stuff

https://github.com/user-attachments/assets/92c79167-cd7e-44a1-8d14-d866b10eb6e0

frenzibyte commented:
@PercyDan54 Cannot reproduce on my iPad. Try debugging through the `AndroidOrientationManager` and trace down what orientation values are being fed to the activity. This looks like when you paused gameplay, `RequestedOrientation` got set to a value and that value caused the game to appear in that broken state.

PercyDan54 commented:
> @PercyDan54 Cannot reproduce on my iPad. Try debugging through the `AndroidOrientationManager` class and trace down what orientation values are being fed to the activity. This looks like when you paused gameplay, `RequestedOrientation` got set to a value and that value caused the game to appear in that broken state.

It's not only when pausing the game once. If I put my tablet in Landscape it immediately enters that broken state when gameplay starts.

The Orientation in gameplay is `FullPortrait`

PercyDan54 commented:
I had a quick look in Android docs and seems the game is put into [Letterboxing mode](https://developer.android.com/guide/practices/device-compatibility-mode#letterboxing)

> On [Android 12 (API level 31)](https://developer.android.com/about/versions/12/12L/summary#dev-device-orientation-request) and higher, device manufacturers can configure individual device screens (such as the tablet-sized screen of a foldable) to ignore this suggestion, and instead force an activity to be letterboxed within the user's preferred orientation of the device.

frenzibyte commented:
Sounds like it's working correctly then.

PercyDan54 commented:
> Sounds like it's working correctly then.

I guess my case might fall into the "device manufacturers configured". Imo it's better to test on more tablets to know whether most manufacturers do this, since it's a bit counter-intuitve

frenzibyte committed:
Fix triangles judgement mispositioned on a miss

Similar to mania's `ArgonJudgementPiece`.

peppy commented:
I'm not sure this approach is going to work well. My original thoughts were that we don't change the OS-level orientation locking and just adjust the ruleset's playfield to render according to a user's preference. As per the reported issues above, I have no idea how solid mobile APIs are to have an *application* decide whether the orientation should be switched on the fly at an OS level, disregarding how the user is holding the device.

frenzibyte commented:
I have considered make the playfield render in a portrait-like mode without changing the OS orientation, but there are multiple things to take into consideration when doing this.
 - The change should be applied to the entirety of `Player`, otherwise the pause screen will be displayed incorrectly.
 - The skin editor has to somehow know the `Player` screen is in portrait mode and *somehow* flip itself to display correctly, otherwise skin editing will be completely broken on iOS / iPads.
 - All game overlays will also have to handle this somehow if the player connects a hardware keyboard and pops in the top toolbar or otherwise.

UIKit define explicit API to let the device know that the orientation needs to be updated, see [`setNeedsUpdateOfSupportedInterfaceOrientations`](https://developer.apple.com/documentation/uikit/uiviewcontroller/setneedsupdateofsupportedinterfaceorientations()?language=objc) (or the old version of this, [`attemptRotationToDeviceOrientation`](https://developer.apple.com/documentation/uikit/uiviewcontroller/attemptrotationtodeviceorientation()?language=objc)).

I'm not sure what are the conditions for making Android switch to letterbox mode as in the reported issue above. Need someone that can investigate that and see how bad it is and/or if letterboxing can be avoided.

frenzibyte committed:
Fix player no longer handling non-loaded beatmaps

frenzibyte committed:
Fix `DrawableManiaRuleset` not cached as itself in subtypes

i.e. editor mania ruleset

Susko3 commented:
> * On tablet
>   * Running osu!mania switches orientation to "portrait". When in gameplay period, orientation is locked, and when paused or in break you can switch orientation between portrait and upside-down portrait.

I (and the mentioned android xiaomi tablet) disagree with this. There is no good reason to force portrait orientation on tablets. People may want to play osu!mania with a keyboard, and they will usually want to have their tablet in landscape mode.

Susko3 commented:
<details><summary>Long screenshot</summary>
<p>

 ![image](https://github.com/user-attachments/assets/c424d500-4d1b-4e51-a67c-a24c1f3bda01)

</p>
</details> 

The last column in mania 7K is blocked by the pause button. Would 7K be more playable in portrait or landscape?

frenzibyte commented:
> > * On tablet
> >   
> >   * Running osu!mania switches orientation to "portrait". When in gameplay period, orientation is locked, and when paused or in break you can switch orientation between portrait and upside-down portrait.
> 
> I (and the mentioned android xiaomi tablet) disagree with this. There is no good reason to force portrait orientation on tablets. People may want to play osu!mania with a keyboard, and they will usually want to have their tablet in landscape mode.

Landscape orientation for mania on tablets can be brought back. It felt unplayable at first with the idea of "touchable columns" and no touch input overlay, but upon testing again it seems roughly fine. However I find it slightly weird to allow landscape orientation when portrait orientation works and works well. I would rather recommend tablet users to always be on portrait if they intend to play mania. @peppy thoughts?

frenzibyte commented:
> Long screenshot
> ![image](https://private-user-images.githubusercontent.com/16479013/400161819-c424d500-4d1b-4e51-a67c-a24c1f3bda01.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYwMjUyMTcsIm5iZiI6MTczNjAyNDkxNywicGF0aCI6Ii8xNjQ3OTAxMy80MDAxNjE4MTktYzQyNGQ1MDAtNGQxYi00ZTUxLWE2N2MtYTI0YzFmM2JkYTAxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA0VDIxMDgzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdmYzM2NTRkMTQ5MzE3N2EyYWEwMWI4MWEyYjUzZDQ4MTRmNmExYWFjYzc3NzM5NWFlYTcwMmU3MTUxYjE0OTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ktEKhDOU1t8cSsPQjd6nWhAz4KqPLsYF8WDj4toUBs8)
> 
> The last column in mania 7K is blocked by the pause button. Would 7K be more playable in portrait or landscape?

1. this is 10K mania
2. that screenshot looks like you're playing on a phone. Landscape will be extremely worse, and 10K mania on phone is practically impossible to have. I doubt it's even playable with touchable columns. It cannot be salvaged in this PR.

peppy commented:
> I would rather recommend tablet users to always be on portrait if they intend to play mania. 

On the flip side, can we not force the orientation change and leave it up to the user to do this? Or does that not work well?

frenzibyte commented:
I wouldn't say it's totally unplayable on landscape orientation on my iPad, but it feels bad with default skins due to the default column width (the column is barely wide enough to cover up my finger). Users will most certainly go through making legacy skins with wider columns to be able to play on it (and I assume it's something we want to expose as a skin setting to support default skins in the future as well).

I can still allow landscape orientation on iPads/tablets. I'll just do that.

frenzibyte committed:
Merge branch 'master' into mobile-fix-mania

frenzibyte committed:
Allow landscape orientation on tablet devices in osu!mania

frenzibyte committed:
Fix failing tests

- Caches `DrawableRuleset` in editor compose screen for mania playfield adjustment container (because it's used to wrap the blueprint container as well)
- Fixes `ManiaModWithPlayfieldCover` performing a no-longer-correct direct cast with a naive-but-working approach.

frenzibyte committed:
Fix mania fade in test not actually testing the mod

peppy committed:
Merge branch 'master' into mobile-fix-mania

peppy commented:
On a read of the code and a test of how this looks in portrait (on PC, but that counts right?) I'm pretty on board with this direction.

[e5713e5](https://github.com/ppy/osu/pull/31368/commits/e5713e52392066a1430ebce460d07d8af01ad29f) caught me off-guard –  is this something you could PR separately along with a visual explanation? I think bundling it with this PR makes things a bit hard to swallow.

peppy commented:
@Susko3 are you able to give this a go/no-go for android (remaining blocking bullet point)?

peppy reviewed (commented):
Initial pass, no major issues.

Susko3 commented:
> @Susko3 are you able to give this a go/no-go for android (remaining blocking bullet point)?

It's good from my side. I've tested as set out in https://github.com/ppy/osu/pull/31368#issuecomment-2566674159. I've only tested on a phone, but I've also simulated a tablet by forcing `IsTablet = true`. Works well in both cases. 4K mania is comfortable to play in portrait.

Tagging @PercyDan54 to give it a test on a real tablet, as things are updated and should work better.

peppy committed:
Mark `IsTablet` with `new` to avoid inspection

Co-authored-by: Susko3 <Susko3@protonmail.com>

frenzibyte commented on osu.Android/OsuGameActivity.cs:
```diff
@@ -49,6 +49,8 @@ public class OsuGameActivity : AndroidGameActivity
         /// <remarks>Adjusted on startup to match expected UX for the current device type (phone/tablet).</remarks>
         public ScreenOrientation DefaultOrientation = ScreenOrientation.Unspecified;
 
+        public bool IsTablet { get; private set; }
```
That sounds like SDL is doing exactly what we're manually doing ourselves? src: https://github.com/revery-ui/esy-sdl2/blob/5f771c6c12a41e2f4295cc1d796b63741fa3c020/android-project/app/src/main/java/org/libsdl/app/SDLActivity.java#L1069-L1075

Not bothering with changing that in here though, can be done in a separate PR.

PercyDan54 commented:
The column is a bit too thin to me in landscape, I'm not a mania player so can't really judge.
![IMG_20250117_133139893](https://github.com/user-attachments/assets/3936ad1e-4a03-4f64-8a77-a8107eb3f27a)

And a small visual bug is the re-layout happens after the countdown which is kinda weird to see the playfield going offscreen for 3 seconds
![Screenshot_2025-01-17-13-30-00-962_sh ppy osulazer](https://github.com/user-attachments/assets/05b080b2-37c4-481b-99df-85311819d325)

frenzibyte commented:
> The column is a bit too thin to me in landscape, I'm not a mania player so can't really judge.

I'll say this is the expectation of attempting to play on landscape, it won't always feel good. Switch to portrait for better experience. Allowing landscape was a choice made few comments above yours https://github.com/ppy/osu/pull/31368#issuecomment-2579150161.

> And a small visual bug is the re-layout happens after the resume countdown which is kinda weird to see the playfield going offscreen for 3 seconds

This is to do with the playfield not being updated while the game is paused, it is out of scope in this PR and most relevant to https://github.com/ppy/osu/issues/7285

PercyDan54 commented:
> > The column is a bit too thin to me in landscape, I'm not a mania player so can't really judge.
> 
> I'll say this is the expectation of attempting to play on landscape, it won't always feel good. Switch to portrait for better experience. Allowing landscape was a choice made few comments above yours [#31368 (comment)](https://github.com/ppy/osu/pull/31368#issuecomment-2579150161).
> 
> > And a small visual bug is the re-layout happens after the resume countdown which is kinda weird to see the playfield going offscreen for 3 seconds
> 
> This is to do with the playfield not being updated while the game is paused, it is out of scope in this PR and most relevant to #7285

Alright.
Portrait 4K looks pretty good to me, until 7K it reaches just larger than the width of my finger. And 8K is around what the above photo (landscape 4K). But I understand that such high key count is not a focus now and I would say the pr is good to go as a first step.

peppy commented:
Given the above responses, I'm going to do a code-level review pass this week with the intention of getting it in 👍 .

frenzibyte committed:
Revert "Fix triangles judgement mispositioned on a miss"

This reverts commit e5713e52392066a1430ebce460d07d8af01ad29f.

frenzibyte commented:
> On a read of the code and a test of how this looks in portrait (on PC, but that counts right?) I'm pretty on board with this direction.
> 
> [e5713e5](https://github.com/ppy/osu/pull/31368/commits/e5713e52392066a1430ebce460d07d8af01ad29f) caught me off-guard – is this something you could PR separately along with a visual explanation? I think bundling it with this PR makes things a bit hard to swallow.

I would have to specifically revert changes to the `DrawableManiaJudgement` class as well, which is part of the https://github.com/ppy/osu/commit/1e08b3dbdac1ed07fd56c0d55d83ce200053c336 commit...I think it will take less time if I briefly explain it here alongside some visual showcasing (especially since merging this without this change will just break triangles skin completely).

This is happening because in  https://github.com/ppy/osu/pull/31368/commits/1e08b3dbdac1ed07fd56c0d55d83ce200053c336 the anchoring of the judgement container has changed so that judgements get moved upwards proportional to the upscaling factor of the playfield.

And due to that change, the positioning of triangles skin mania judgements is completely broken as they're used to display close to the top of the playfield. Therefore both the `DrawableManiaJudgement` part of https://github.com/ppy/osu/commit/1e08b3dbdac1ed07fd56c0d55d83ce200053c336 as well as https://github.com/ppy/osu/pull/31368/commits/e5713e52392066a1430ebce460d07d8af01ad29f change the positioning of said judgements to become similar to argon and work correctly.

Before:

https://github.com/user-attachments/assets/d78b74ad-bdf3-49b2-a45f-cb37c9ec76d3

After:

https://github.com/user-attachments/assets/d6a3db1d-8ffd-4b20-b110-b4552e1835e3

frenzibyte committed:
Reapply "Fix triangles judgement mispositioned on a miss"

This reverts commit 6ec718304e4df307d8ae3598de96585ff836a99e.

peppy commented on osu.Game.Rulesets.Mania/UI/ManiaPlayfieldAdjustmentContainer.cs:
```diff
@@ -1,17 +1,66 @@
+        }
+
+        [Resolved]
+        private DrawableRuleset drawableRuleset { get; set; } = null!;
```
Let's not:

```diff
diff --git a/osu.Game.Rulesets.Mania/UI/DrawableManiaRuleset.cs b/osu.Game.Rulesets.Mania/UI/DrawableManiaRuleset.cs
index a186d9aa7d..e33cf092c3 100644
--- a/osu.Game.Rulesets.Mania/UI/DrawableManiaRuleset.cs
+++ b/osu.Game.Rulesets.Mania/UI/DrawableManiaRuleset.cs
@@ -161,7 +161,7 @@ private void updateTimeRange()
         /// <returns>The scroll time.</returns>
         public static double ComputeScrollTime(double scrollSpeed) => MAX_TIME_RANGE / scrollSpeed;
 
-        public override PlayfieldAdjustmentContainer CreatePlayfieldAdjustmentContainer() => new ManiaPlayfieldAdjustmentContainer();
+        public override PlayfieldAdjustmentContainer CreatePlayfieldAdjustmentContainer() => new ManiaPlayfieldAdjustmentContainer(this);
 
         protected override Playfield CreatePlayfield() => new ManiaPlayfield(Beatmap.Stages);
 
diff --git a/osu.Game.Rulesets.Mania/UI/ManiaPlayfieldAdjustmentContainer.cs b/osu.Game.Rulesets.Mania/UI/ManiaPlayfieldAdjustmentContainer.cs
index b0203643b0..feb75b9f1e 100644
--- a/osu.Game.Rulesets.Mania/UI/ManiaPlayfieldAdjustmentContainer.cs
+++ b/osu.Game.Rulesets.Mania/UI/ManiaPlayfieldAdjustmentContainer.cs
@@ -2,7 +2,6 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System;
-using osu.Framework.Allocation;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Containers;
 using osu.Game.Rulesets.UI;
@@ -16,8 +15,11 @@ public partial class ManiaPlayfieldAdjustmentContainer : PlayfieldAdjustmentCont
 
         private readonly DrawSizePreservingFillContainer scalingContainer;
 
-        public ManiaPlayfieldAdjustmentContainer()
+        private readonly DrawableManiaRuleset drawableManiaRuleset;
+
+        public ManiaPlayfieldAdjustmentContainer(DrawableManiaRuleset drawableManiaRuleset)
         {
+            this.drawableManiaRuleset = drawableManiaRuleset;
             InternalChild = scalingContainer = new DrawSizePreservingFillContainer
             {
                 Anchor = Anchor.Centre,
@@ -30,9 +32,6 @@ public ManiaPlayfieldAdjustmentContainer()
             };
         }
 
-        [Resolved]
-        private DrawableRuleset drawableRuleset { get; set; } = null!;
-
         protected override void Update()
         {
             base.Update();
@@ -40,8 +39,6 @@ protected override void Update()
             float aspectRatio = DrawWidth / DrawHeight;
             bool isPortrait = aspectRatio < 1f;
 
-            var drawableManiaRuleset = (DrawableManiaRuleset)drawableRuleset;
-
             if (isPortrait && drawableManiaRuleset.Beatmap.Stages.Count == 1)
             {
                 // Scale playfield up by 25% to become playable on mobile devices,
diff --git a/osu.Game/Rulesets/Edit/DrawableEditorRulesetWrapper.cs b/osu.Game/Rulesets/Edit/DrawableEditorRulesetWrapper.cs
index 573eb8c42f..174b278d89 100644
--- a/osu.Game/Rulesets/Edit/DrawableEditorRulesetWrapper.cs
+++ b/osu.Game/Rulesets/Edit/DrawableEditorRulesetWrapper.cs
@@ -19,16 +19,16 @@ namespace osu.Game.Rulesets.Edit
     internal partial class DrawableEditorRulesetWrapper<TObject> : CompositeDrawable
         where TObject : HitObject
     {
-        public Playfield Playfield => DrawableRuleset.Playfield;
+        public Playfield Playfield => drawableRuleset.Playfield;
 
-        public readonly DrawableRuleset<TObject> DrawableRuleset;
+        private readonly DrawableRuleset<TObject> drawableRuleset;
 
         [Resolved]
         private EditorBeatmap beatmap { get; set; } = null!;
 
         public DrawableEditorRulesetWrapper(DrawableRuleset<TObject> drawableRuleset)
         {
-            DrawableRuleset = drawableRuleset;
+            this.drawableRuleset = drawableRuleset;
 
             RelativeSizeAxes = Axes.Both;
 
@@ -38,7 +38,7 @@ public DrawableEditorRulesetWrapper(DrawableRuleset<TObject> drawableRuleset)
         [BackgroundDependencyLoader]
         private void load()
         {
-            DrawableRuleset.FrameStablePlayback = false;
+            drawableRuleset.FrameStablePlayback = false;
             Playfield.DisplayJudgements.Value = false;
         }
 
@@ -67,27 +67,27 @@ protected override void LoadComplete()
 
         private void regenerateAutoplay()
         {
-            var autoplayMod = DrawableRuleset.Mods.OfType<ModAutoplay>().Single();
-            DrawableRuleset.SetReplayScore(autoplayMod.CreateScoreFromReplayData(DrawableRuleset.Beatmap, DrawableRuleset.Mods));
+            var autoplayMod = drawableRuleset.Mods.OfType<ModAutoplay>().Single();
+            drawableRuleset.SetReplayScore(autoplayMod.CreateScoreFromReplayData(drawableRuleset.Beatmap, drawableRuleset.Mods));
         }
 
         private void addHitObject(HitObject hitObject)
         {
-            DrawableRuleset.AddHitObject((TObject)hitObject);
-            DrawableRuleset.Playfield.PostProcess();
+            drawableRuleset.AddHitObject((TObject)hitObject);
+            drawableRuleset.Playfield.PostProcess();
         }
 
         private void removeHitObject(HitObject hitObject)
         {
-            DrawableRuleset.RemoveHitObject((TObject)hitObject);
-            DrawableRuleset.Playfield.PostProcess();
+            drawableRuleset.RemoveHitObject((TObject)hitObject);
+            drawableRuleset.Playfield.PostProcess();
         }
 
         public override bool PropagatePositionalInputSubTree => false;
 
         public override bool PropagateNonPositionalInputSubTree => false;
 
-        public PlayfieldAdjustmentContainer CreatePlayfieldAdjustmentContainer() => DrawableRuleset.CreatePlayfieldAdjustmentContainer();
+        public PlayfieldAdjustmentContainer CreatePlayfieldAdjustmentContainer() => drawableRuleset.CreatePlayfieldAdjustmentContainer();
 
         protected override void Dispose(bool isDisposing)
         {
diff --git a/osu.Game/Rulesets/Edit/HitObjectComposer.cs b/osu.Game/Rulesets/Edit/HitObjectComposer.cs
index 8882d55b42..15b60114af 100644
--- a/osu.Game/Rulesets/Edit/HitObjectComposer.cs
+++ b/osu.Game/Rulesets/Edit/HitObjectComposer.cs
@@ -133,7 +133,6 @@ private void load(OsuConfigManager config, [CanBeNull] Editor editor)
             if (DrawableRuleset is IDrawableScrollingRuleset scrollingRuleset)
                 dependencies.CacheAs(scrollingRuleset.ScrollingInfo);
 
-            dependencies.CacheAs<DrawableRuleset>(drawableRulesetWrapper.DrawableRuleset);
             dependencies.CacheAs(Playfield);
 
             InternalChildren = new[]

```

frenzibyte replied:
https://github.com/ppy/osu/pull/31368/commits/ffc37cece0483c9bcdea0962abc8bfbe1dd9b0f1

peppy reviewed (changes requested):
Just a code level check

peppy committed:
Merge branch 'master' into mobile-fix-mania

frenzibyte commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -68,6 +68,8 @@ public abstract partial class Player : ScreenWithBeatmapBackground, ISamplePlayb
 
         public override bool HideMenuCursorOnNonMouseInput => true;
 
+        public override bool RequiresPortraitOrientation => DrawableRuleset?.RequiresPortraitOrientation == true;
```
Sounds better, I was going to write it this way but felt like being too verbose for what the meaning of a null `DrawableRuleset` is supposed to be...maybe not. Will apply.

frenzibyte replied:
https://github.com/ppy/osu/pull/31368/commits/dac7d21302cbd9b7094ba7fc0d5989a9f254d46d

frenzibyte committed:
Be explicit on nullability in `RequiresPortraitOrientation`

Co-authored-by: Dean Herbert <pe@ppy.sh>

frenzibyte commented on osu.Game.Rulesets.Mania/UI/Components/HitPositionPaddedContainer.cs:
```diff
@@ -1,29 +1,32 @@
+            InternalChild = child;
+        }
+
+        internal void Add(Drawable drawable)
```
Technically not, since all that `SkinReloadableDrawable` does is trigger an event when the selected skin is changed. Now that I look at it, I can totally avoid this inheritance by just...resolving the skin source and subscribing to skin change...

frenzibyte replied:
https://github.com/ppy/osu/pull/31368/commits/8151c3095ddfc6389516054c4ae66ead80f5b605

frenzibyte committed:
Revert unnecessary inheritance

Everyone is right, too much inheritance and polymorphism backfires very badly.

frenzibyte committed:
Avoid extra unnecessary DI

Co-authored-by: Dean Herbert <pe@ppy.sh>

frenzibyte committed:
Simplify orientation locking code magnificently

frenzibyte committed:
Remove `OrientationManager` and the entire mobile namespace

frenzibyte commented on osu.Game/Mobile/OrientationManager.cs:
```diff
@@ -0,0 +1,85 @@
+using osu.Framework.Graphics;
+using osu.Game.Screens.Play;
+
+namespace osu.Game.Mobile
```
Good call, sounds reasonable. I've removed `OrientationManager` entirely and moved any shared logic to a `MobileUtils` static class (name is arbitrary, can be changed, can be renamed to `OrientationUtils` /shrug). https://github.com/ppy/osu/pull/31368/commits/c18128e97419ea1f7c9a4086f1b19de8f9c6022e

frenzibyte commented on osu.Game/Mobile/OrientationManager.cs:
```diff
@@ -0,0 +1,85 @@
+            }
+
+            SetAllowedOrientations(null);
+        }
```
Indeed, it is quite easy to simplify this since we changed tablet devices to not be affected by the "portrait lock" thingy. https://github.com/ppy/osu/pull/31368/commits/bb7daae08063fb06e16934b7542a14b65a1f189d

peppy commented on osu.Game.Rulesets.Mania/UI/DrawableManiaJudgement.cs:
```diff
@@ -32,8 +36,20 @@ public override void PlayAnimation()
+
                     case HitResult.Miss:
-                        base.PlayAnimation();
+                        this.ScaleTo(1.6f);
```
Can you explain the relevance of this code added here?

frenzibyte replied:
Because I've changed the position of the triangles judgement, I have to inline the code in `base.PlayAnimation` and specify the Y position to be `judgement_y_position` instead of zero like in base.

peppy committed:
Fix skin `SourceChanged` event never being unbound

peppy commented on osu.Game.Rulesets.Mania/UI/Column.cs:
```diff
@@ -180,5 +184,29 @@ public void OnReleased(KeyBindingReleaseEvent<ManiaAction> e)
+
+        #region Touch Input
+
+        [Resolved(canBeNull: true)]
+        private ManiaInputManager maniaInputManager { get; set; }
```
can be null?

frenzibyte replied:
Can be null according to some mania editor tests that initialise `Column`s without surrounding input managers. I've enabled NRT to fix usages bypassing this.

peppy commented on osu.Game.Rulesets.Mania/UI/Column.cs:
```diff
@@ -180,5 +184,29 @@ public void OnReleased(KeyBindingReleaseEvent<ManiaAction> e)
+
+        protected override bool OnTouchDown(TouchDownEvent e)
+        {
+            maniaInputManager.KeyBindingContainer.TriggerPressed(Action.Value);
```
or not null?

frenzibyte committed:
Enable NRT on `Column`

peppy committed:
Remove pointless comment

X-X-X-X-X-X-X-X-X-X-X-X-X commented:
![73ea99b6f9730b06c1a612a5d84fc545](https://github.com/user-attachments/assets/b617ffa2-477d-4df5-9f9b-1270488d9011)
The vertical screen experience is decent, but not many people use it to play Mania. The horizontal buttons are still too narrow, and perhaps people would prefer the horizontal screen to fill the entire screen. once https://github.com/ppy/osu/releases/tag/2024.412.1 This version is actually quite good for mobile phones
